### PR TITLE
chore: Add type util

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    MISC: 导出 `GetProps`、`GetProp`、`GetRef` 工具方法，便于开发者获取未导出的类型定义。       |
| 🇨🇳 Chinese |   MISC: Export `GetProps` & `GetProp` & `GetRef` type util for developer to get the definition that antd not exported.        |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Review by li-jia-nan — APPROVED** [review]

File: components/_util/type.ts [hunk 1]
```diff
@@ -1,6 +1,32 @@
...
+  : T extends object
+    ? T
+    : never;
+
+export type GetProp<
+  T extends React.ComponentType<any> | object,
+  PropName extends keyof GetProps<T>,
+> = NonNullable<GetProps<T>[PropName]>;
+
+type ReactRefComponent<Props extends { ref?: React.Ref<any> }> = (
```
Comment (li-jia-nan) [inline components/_util/type.ts:23]:
```suggestion
type ReactRefComponent<Props extends React.RefAttributes<any>> = (
```

**Review by kiner-tang — APPROVED** [review]

File: components/_util/type.ts [hunk 2]
```diff
@@ -1,6 +1,32 @@
+import type React from 'react';
+
 /** https://github.com/Microsoft/TypeScript/issues/29729 */
 export type LiteralUnion<T extends string> = T | (string & {});
 
 export type AnyObject = Record<PropertyKey, any>;
 
 export type CustomComponent<P = AnyObject> = React.ComponentType<P> | string;
+
+export type GetProps<T extends React.ComponentType<any> | object> = T extends React.ComponentType<
```
Comment (li-jia-nan) [inline components/_util/type.ts:10]: 这个 object 可以换成 AnyObject 吗？

File: components/_util/type.ts (hunk 1 above)
Comment (zombieJ) [reply components/_util/type.ts:23]: 这个估计搞成最简类型的哈~

Comment (li-jia-nan) [reply components/_util/type.ts:23]: soga

File: components/_util/type.ts (hunk 2 above)
Comment (zombieJ) [reply components/_util/type.ts:10]: 体操类型不要使用 antd 内部的定义，否则外面会传染。它的干涉类型越简单越好哈~

Comment (li-jia-nan) [reply components/_util/type.ts:10]: 学到了

File: components/float-button/demo/badge-debug.tsx [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import React, { useState } from 'react';
 import { ConfigProvider, FloatButton, Slider } from 'antd';
-import type { AliasToken } from 'antd/es/theme/interface';
+import type { ConfigProviderProps, GetProp } from 'antd';
+
+type AliasToken = GetProp<ConfigProviderProps, 'theme'>['token'];
```
Comment (Wxh16144) [inline components/float-button/demo/badge-debug.tsx:5]: 这和直接写 `ConfigProviderProps['theme']['token']` 好像没区别🤔

Comment (zombieJ) [reply components/float-button/demo/badge-debug.tsx:5]: `GetProp` 会加个非空转换，直接这么写是不会过的。

**Comment by li-jia-nan** [comment]
改动有点多，感觉还得 merge branch 一下，不然后面冲突一大堆

**Pull request merged**